yarn: Fix state=latest not working with global=true

sargunv opened the pull request:
##### SUMMARY

Fixes #5712.

* Updates `_exec` to support running Yarn commands that wouldn't ordinarily be available with `yarn global` by setting the working directory to the path from `yarn global dir` instead.
* Updates `list_outdated` to use this new functionality when invoking `yarn outdated`.
* Updates `list_outdated` to check for actual errors in stderr output instead of failing on the presence of anything in stderr. This is b/c the Yarn global package.json has no license field, causing a warning to be logged to stderr when invoking `yarn outdated` in that directory.

Through testing, found a couple more related bugs to fix (see comments below):
* Replaces the hardcoded `DEFAULT_GLOBAL_INSTALLATION_PATH` with code to check the output of `yarn global dir`, as the default path may be different when the use is root, or when the user has set YARN_GLOBAL_FOLDER or the corresponding config flag.
* Updates `list` to function correctly for packages without binaries (regular libraries) when global=true by invoking `yarn list` without `global`, the same way I did for `yarn outdated` above.

And finally, added tests for all the above cases. Previously, there were no tests for `global=true` at all.

##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME

yarn

##### ADDITIONAL INFORMATION

Before:

![image](https://user-images.githubusercontent.com/1320357/212242061-8185be66-5083-496a-b0e0-f97177f518ab.png)

After:

![image](https://user-images.githubusercontent.com/1320357/212242095-5f6f5135-16c7-4505-8466-f330e84fcfb5.png)

sargunv commented on plugins/modules/yarn.py:
```diff
@@ -206,16 +208,23 @@ def _exec(self, args, run_in_check_mode=False, check_rc=True):
-            if self.path and not self.globally:
-                if not os.path.exists(self.path):
+
+            if self.globally and unsupported_with_global:
+                _rc, out, _err = self.module.run_command(self.executable + ['global', 'dir'], check_rc=check_rc)
+                path = out.strip()
```
This does mean a `global=true` param will override the `path` param, but that was previously true anyway (since `yarn global` from any working directory would operate on the global packages).

russoz commented:
Hi @sargunv 
Thank you for the contribution! Please add a [changelog fragment](https://docs.ansible.com/ansible/devel/community/development_process.html#changelogs-how-to).

Also, if you could please add a testcase in the integration tests, so future modifications in this module take that change into account.

sargunv commented:
There were previously no `yarn global` tests, so I added test cases to global install, upgrade, and remove a package.

The test for upgrade (state=latest, what this PR fixes) is passing, but the test for remove (state=absent) is failing. I'll have to look into how to get these tests running locally to investigate

sargunv commented:
Got the tests working locally (turns out WSL can enable systemd) and found the issue with the `yarn global remove` test.

The yarn module assumes a certain `yarn global dir` but this assumption is incorrect when running as the root user (as the tests do). When root, the default `yarn global dir` appears to be `/usr/local/share/.config/yarn/global/`, not `/root/.config/yarn/global/`

https://github.com/ansible-collections/community.general/blob/44172ddaa6130d2441954da54da0ea70caf41521/plugins/modules/yarn.py#L166

This is a totally different bug from what this PR addresses so I'll file that fix as its own PR, and add that test there.

russoz commented:
The magic of tests, we find problems we did not imagine we had 😁.

Thanks!

sargunv commented:
Fixed that, and while testing _that_ fix I found yet another little bug 🙃 

The `list()` function uses `yarn list` or `yarn global list` depending on whether global=true, but `yarn global list` only lists packages with binaries, not all globally installed packages. So, cases like `name=left-pad state=absent global=true` do not work; the step thinks `left-pad` is already absent.

I updated that one to use the same `unsupported_with_global` param I added for the bug this PR was initially addressing, and removed the bit of code that handled the different output format of `yarn global list`, and added tests for this case too.

felixfontein commented on changelogs/fragments/5829-fix-yarn-global.yml:
```diff
@@ -0,0 +1,4 @@
+bugfixes:
+  - yarn - fix state=latest not working with global=true (https://github.com/ansible-collections/community.general/pull/5829).
+  - yarn - fix global=true to check for the configured global folder instead of assuming the default (https://github.com/ansible-collections/community.general/pull/5829)
+  - yarn - fix state=absent not working with global=true when the package does not include a binary (https://github.com/ansible-collections/community.general/pull/5829)
```
```suggestion
  - yarn - fix ``state=latest`` not working with ``global=true`` (https://github.com/ansible-collections/community.general/pull/5829).
  - yarn - fix ``global=true`` to check for the configured global folder instead of assuming the default (https://github.com/ansible-collections/community.general/pull/5829)
  - yarn - fix ``state=absent`` not working with ``global=true`` when the package does not include a binary (https://github.com/ansible-collections/community.general/pull/5829)
```

sargunv commented:
Hi, is there anything else I need to do to make this PR ready to merge?

russoz commented on changelogs/fragments/5829-fix-yarn-global.yml:
```diff
@@ -0,0 +1,4 @@
+bugfixes:
+  - yarn - fix ``state=latest`` not working with ``global=true`` (https://github.com/ansible-collections/community.general/pull/5829).
```
```suggestion
  - yarn - fix ``state=latest`` not working with ``global=true`` (https://github.com/ansible-collections/community.general/issues/5712).
```

russoz replied:
When fixing a reported bug, please refer to it. Alternatively, you may list both of them separated by comma - there is no specific norm about doing one way or the other.

russoz reviewed (commented):
Hi @sargunv it looks good to me. Only one small issue with the changelog fragment.

sargunv commented:
Fixed!

russoz reviewed (approved):
LGTM

felixfontein commented:
@sargunv thanks for fixing this!
@russoz thanks for reviewing!
